### ISSUE TITLE
Need help to install electroncash on Linux (Tails )

### ISSUE THREAD

**Issue by DebiantDeviant** [fyookball/electrum#393]
Hello, 

I need some help please to install Electroncash on Tails if possible; Just to let you know i am fairly new with using a Linux system so bare with me if i dont follow what you are asking me.
First, 
I already downloaded and verified the signature of the .tar înstall file for Linux located on electoncash.org

My question is simple i think, What is the command line to start the install process if i execute it from the terminal window in the Electron Cash-3.0 folder??  Do i need specific dependecies installed on my system before the installation?? If yes, how can i verify if they are installed on my system? And where can i download the missing files or tools?

I'm asking this because on my version of Tails, 3.3, i had to update the Electrum version and the only way i could start the program was thru the terminal window using the following command line: 'python3 electrum' , the command line 'sudo python3 setup.py install' did not work so i guess i might have to use the same process to open Electroncash scince it did not come pre-built in the current version of Tails, ( i read it might be in the next built. i hope so!!)
So if you could help me with this basic info it would be much appreciated, 
thanx!

**Comment by 1ib3r7yr3igns** [comment]
I'd like to upvote this. I downloaded ElectronCash3.0.tar.gz and I was unable to open it in Tails version 3.3. In Electron Cash version 2.9.4, I would just run
`python electron-cash`,

But now, when I do,

`python3 electron-cash`, 

or 

`python electron-cash`

or 

`./electron-cash`

I get missing python dependencies.

setup.py install did not work either, though I'm not so sure installation is any good on Tails since it'll be wiped anyway after reboot.

**Comment by QuavoLecso** [comment]
I can't install it either,

**Comment by marceloneil** [comment]
Here is how you install Electron Cash 3.0 on linux:

First retrieve @fyookball's public key from his repo, you only need to do this once but it is recommended to insure that you are installing an un-altered version of electron cash.
```bash
wget "https://github.com/fyookball/keys-n-hashes/raw/master/pubkeys/jonaldkey2.txt"
gpg --import jonaldkey2.txt
rm jonaldkey2.txt
```
Next download the actual application, signature and checksum
```bash
wget "https://github.com/fyookball/electrum/releases/download/3.0/ElectronCash-3.0.tar.gz"
wget "https://github.com/fyookball/keys-n-hashes/raw/master/sigs-and-sums/3.0/win-linux/ElectronCash-3.0.tar.gz.sig"
wget "https://github.com/fyookball/keys-n-hashes/raw/master/sigs-and-sums/3.0/win-linux/SHA256.ElectronCash-3.0.tar.gz.txt"
```
Now verify the download
```bash
sha256sum -c SHA256.ElectronCash-3.0.tar.gz.txt
gpg --verify ElectronCash-3.0.tar.gz.sig
```
Finally install it
```bash
tar -xvzf ElectronCash-3.0.tar.gz
cd Electron\ Cash-3.0/
# Note: you may or may not need to run this as root/sudo
sudo python3 setup.py install
```
Hopefully, you will be able to run electron cash using the `electron-cash` command.

The `setup.py install` command *should* install all of the dependencies you need, but if it doesn't work please provide an output so we can see whats wrong

**Comment by 1ib3r7yr3igns** [comment]
I tried an install on fresh copies of both tails and linux mint. On both I got the same issue when running
`sudo python3 setup.py install`.

Python could not find module 'setuptools'. In order to get setuptools, I needed to install pip3. 

On linux mint, I installed pip3, installed setuptools, then was able to get it all started.

On Tails, I had to run `sudo apt-get update && sudo apt-get upgrade` before I could even install pip3. Once I had pip3 installed, I tried `sudo pip3 install setuptools`, however, it errored out and I was not able to get the setuptools installed, which prevented me from running setup.py.

Since tails is amnesiac, even if I did get setuptools installed, all of it would be wiped after reboot anyway. You'd have to update and upgrade apt-get, install pip3 and install setuptools every time Tails boots up. And since tails only runs over tor, this takes much longer than on a regular linux distro.

**Comment by Blueyre2** [comment]
For some reason, electron-cash-2.9.4 works perfectly fine on Tails without the need of installing any dependencies. I just run ./electron-cash and it works. However, electron-cash-3.0, the latest version, doesn't work on Tails and gives a missing dependency error when attempting to run it by using ./electron-cash. I realized that 2.9.4 comes with the packages folder mentioned in the README and 3.0 doesn't so if you want  to run electron-cash on Tails for the time being, use 2.9.4 and it should work for you.

**Comment by 2600box** [comment]
jumping on this issue because I think it is the same as I have. 

after running `python3 setup.py install` and then `./electron-cash` I get the following missing dependency:

`Error: No module named 'google.protobuf'. Try 'sudo pip install <module-name>'`

which I cannot find a way to satisfy. 

`sudo pip3 install google.protobuf
Collecting google.protobuf
  Could not find a version that satisfies the requirement google.protobuf (from versions: )
No matching distribution found for google.protobuf
`
and I have tried installing protobuf individually with pip3 and with apt with no success.

this is on debian jessie

**Comment by marceloneil** [comment]
The previous releases included a folder called `packages/`, which contained all the dependencies needed, unfortunately the 3.0 build is missing this folder. @fyookball, would you be able to update the linux package and ensure you run `./contrib/make_packages`? It seems to be the root of a lot of these tails users problems.

@1ib3r7yr3igns @Blueyre2 you can compile your own package temporarily on another computer using this code:
```bash
# Download the 3.0 source code
wget "https://github.com/fyookball/electrum/archive/3.0.tar.gz"
tar -xzvf 3.0.tar.gz
cd electrum-3.0/
# Package the dependencies
./contrib/make_packages
# Compile the icons file for Qt
sudo apt-get install pyqt5-dev-tools
pyrcc5 icons.qrc -o gui/qt/icons_rc.py
# Compile the protobuf description file
sudo apt-get install protobuf-compiler
protoc --proto_path=lib/ --python_out=lib/ lib/paymentrequest.proto
# Create translations (optional)
sudo apt-get install python-pycurl gettext
./contrib/make_locale
# Package binary
sudo apt-get install python3-setuptools
python3 setup.py sdist
```
The tar should be in the `dist/` folder

@2600box try running `pip3 install protobuf`, or if you update to stretch you can install `python3-protobuf`

Let me know if there are any more issues

**Comment by marceloneil** [comment]
Just out of curiousity, if tails wipes your data on reboot, where do you store your private keys? A usb key? Or does tails preserve your home directory

**Comment by 2600box** [comment]
@marceloneil, thanks for the reply. 

I have tried these versions of protobuf, but electroncash still complains about google.protobuf specifically.

oh and i said debian jessie out of habit. I meant that i am on stretch.

**Comment by marceloneil** [comment]
@2600box That's quite odd. I've tested it on a vm of debian stretch and [`protobuf==3.5.1`](https://pypi.python.org/pypi/protobuf/3.5.1) works great. Using `python3 setup.py install` downloaded all the dependencies for me (including `protobuf`), so I had to `pip3 uninstall protobuf` to replicate your issue.

**Comment by 1ib3r7yr3igns** [comment]
Thanks marceloneil!

That worked on Linux Mint, and then I copied it over and tried to start it up on Tails, however after ```./electron-cash``` I got the missing python dependency 'dns'.

As to your question, Tails has optional encrypted persistence. There is a folder located at ~/Persistence/ that will persist necessary data stored encrypted. There is also the option to persist ssh keys and gpg keys and even the electrum wallet for BTC (hopefully electron cash soon). But with gpg keys, ssh keys persisted, it is trivial to just restore the wallet from seed each time you boot into tails. As I have it setup now, the wallet keys do not survive reboot (since ~/.electron-cash/ is not persisted).

**Comment by marceloneil** [comment]
Sorry @1ib3r7yr3igns, see my revised post - I forgot to `cp -r contrib/packages .`

**Comment by kyuupichan** [comment]
I'm closing this - this is just a dependency issue when running from source.  You need to know how to install the dependencies; this is not something Electron Cash can help you solve in general - it's specific to your system.

**Comment by marceloneil** [comment]
This has been fixed in the 3.1.2 release
